service/storage2/payload/payload.go

```diff
@@ -0,0 +1,105 @@
+package payload
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/pebble"
+
+	"github.com/onflow/flow-archive/service/storage2"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const (
+	// Size of the block height encoded in the key.
+	heightSize = 8
+)
+
+type payloadStorage struct {
+	db *pebble.DB
+}
+
+// NewPayloadStorage creates a pebble-backed payload storage.
+// The reason we use a separate storage for payloads we need a Comparer with a custom Split function.
+//
+// It needs to access the last available payload with height less or equal to the requested height.
+// This means all point-lookups are range scans.
+func NewStorage(dbPath string, cacheSize int64) (*payloadStorage, error) {
+	// A custom Split function splits the key into the prefix and the suffix.
+	// This is needed for SeekPrefixGE to work.
+	comparer := *pebble.DefaultComparer
+	comparer.Split = func(a []byte) int {
+		return len(a) - heightSize
+	}
+	comparer.Name = "flow.PlayloadHeightComparer"
+
+	// TODO(rbtz): cache metrics
+	cache := pebble.NewCache(cacheSize)
+	defer cache.Unref()
+
+	opts := storage2.DefaultPebbleOptions(cache, &comparer)
+	db, err := pebble.Open(dbPath, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open db: %w", err)
+	}
+
+	return &payloadStorage{
+		db: db,
+	}, nil
+}
+
+// GetPayload returns the payload for the given height and RegisterID.
+func (s *payloadStorage) GetPayload(
+	height uint64,
+	reg flow.RegisterID,
+) ([]byte, error) {
+	iter := s.db.NewIter(&pebble.IterOptions{
+		UseL6Filters: true,
+	})
+	defer iter.Close()
+
+	encoded := newLookupKey(height, reg).Bytes()
+	ok := iter.SeekPrefixGE(encoded)
+	if !ok {
+		return nil, &ErrNotFound{
```

Comment:
in general, there's no such thing as not found in flow.  not found implies empty string

Reply:
Yeah, this makes sense in merkle trie, not sure whether we need to carry the same assumption to the database-baked register storage.  Let me sleep on it.

Reply:
Yeah, you are right.  By doing this we'll be able to simplify this codepath: https://github.com/onflow/flow-archive/blob/6f8333901d5539e7d5c6f84e86d20f7f80dfbf27/service/index/reader.go#L120-L125

Reply:
This actually has nothing to do with merkle tree in general.  This is a flow specific design choice that we are stuck with (having not found error in the merkle tree would have actually make things easier).  Returning not found error is actually a bug.